File: q0152/Q0152.py
def maxProduct(nums):
    res, curr_max, curr_min = nums[0], nums[0], nums[0]
    for num in nums[1:]:
        tempMax = max(num, curr_max * num, curr_min * num)
        tempMin = min(num, curr_max * num, curr_min * num)
        curr_max = tempMax
        curr_min = tempMin
        res = max(res, curr_max)
    return res

File: q0152/q0152.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int currMax = nums[0];
        int currMin = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int currNum = nums[i];
            int newCurrMax = currMax * currNum;
            int newCurrMin = currMin * currNum;
            currMax = Math.max(newCurrMax, Math.max(newCurrMin, currNum));
            currMin = Math.min(newCurrMax, Math.min(newCurrMin, currNum));
            res = Math.max(res, currMax);
        }
        return res;
    }
}